// @observablehq/framework/dist/render.js
import mime from "mime";
import { mergeToc } from "./config.js";
import { getClientPath } from "./files.js";
import { html, parseHtml, rewriteHtml } from "./html.js";
import { transpileJavaScript } from "./javascript/transpile.js";
import { findLink, normalizePath } from "./pager.js";
import { relativePath } from "./path.js";
import { getResolvers } from "./resolvers.js";
import { rollupClient } from "./rollup.js";
async function renderPage(page, options) {
  const { data } = page;
  const { root, md, base, path, pages, title, preview, search, resolvers = await getResolvers(page, options) } = options;
  const { normalizeLink } = md;
  const sidebar = data?.sidebar !== void 0 ? Boolean(data.sidebar) : options.sidebar;
  const toc = mergeToc(data?.toc, options.toc);
  const draft = Boolean(data?.draft);
  const { files, resolveFile, resolveImport } = resolvers;
  return String(html`<!DOCTYPE html>
<meta charset="utf-8">${path === "/404" ? html`\n<base href="${preview ? "/" : base}">` : ""}
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
${page.title || title ? html`<title>${[page.title, page.title === title ? null : title].filter((title2) => !!title2).join(" | ")}</title>\n` : ""}${renderHead(page, resolvers, options)}${path === "/404" ? html.unsafe(`
<script type="module">

if (location.pathname.endsWith("/")) {
  const alt = location.pathname.slice(0, -1);
  fetch(alt, {method: "HEAD"}).then((response) => response.ok && location.replace(alt + location.search + location.hash));
}

</script>`) : ""}
<script type="module">${html.unsafe(`

import ${preview || page.code.length ? `{${preview ? "open, " : ""}define} from ` : ""}${JSON.stringify(
    resolveImport("observablehq:client")
  )};${files.size || data?.sql ? `
import {registerFile${data?.sql ? ", FileAttachment" : ""}} from ${JSON.stringify(
    resolveImport("observablehq:stdlib")
  )};` : ""}${data?.sql ? `
import {registerTable} from ${JSON.stringify(resolveImport("npm:@observablehq/duckdb"))};` : ""}${files.size ? `
${renderFiles(files, resolveFile)}` : ""}${data?.sql ? `
${Object.entries(data.sql).map(([name, source]) => `registerTable(${JSON.stringify(name)}, FileAttachment(${JSON.stringify(source)}));`).join("\n")}` : ""}
${preview ? `
open({hash: ${JSON.stringify(resolvers.hash)}, eval: (body) => eval(body)});
` : ""}${page.code.map(({ node, id }) => `
${transpileJavaScript(node, { id, resolveImport })}`).join("")}`)}
</script>${sidebar ? html`\n${await renderSidebar(title, pages, root, path, search, normalizeLink)}` : ""}${toc.show ? html`\n${renderToc(findHeaders(page), toc.label)}` : ""}
<div id="observablehq-center">${renderHeader(options, data)}
<main id="observablehq-main" class="observablehq${draft ? " observablehq--draft" : ""}">
${html.unsafe(rewriteHtml(page.html, resolvers.resolveFile))}</main>${renderFooter(path, options, data, normalizeLink)}
</div>
`);
}
function renderFiles(files, resolve) {
  return Array.from(files).sort().map((f) => renderFile(f, resolve)).join("");
}
function renderFile(name, resolve) {
  return `
registerFile(${JSON.stringify(name)}, ${JSON.stringify({
    name,
    mimeType: mime.getType(name) ?? void 0,
    path: resolve(name)
  })});`;
}
async function renderSidebar(title = "Home", pages, root, path, search, normalizeLink) {
  return html`<input id="observablehq-sidebar-toggle" type="checkbox" title="Toggle sidebar">
<label id="observablehq-sidebar-backdrop" for="observablehq-sidebar-toggle"></label>
<nav id="observablehq-sidebar">
  <ol>
    <label id="observablehq-sidebar-close" for="observablehq-sidebar-toggle"></label>
    <li class="observablehq-link${normalizePath(path) === "/index" ? " observablehq-link-active" : ""}"><a href="${normalizeLink(relativePath(path, "/"))}">${title}</a></li>
  </ol>${search ? html`\n  <div id="observablehq-search"><input type="search" placeholder="Search"></div>
  <div id="observablehq-search-results"></div>
  <script>{${html.unsafe(
    (await rollupClient(getClientPath("search-init.js"), root, path, { minify: true })).trim()
  )}}</script>` : ""}
  <ol>${pages.map(
    (p, i) => "pages" in p ? html`${i > 0 && "path" in pages[i - 1] ? html`</ol>` : ""}
    <details${p.pages.some((p2) => normalizePath(p2.path) === path) ? html` open class="observablehq-section-active"` : p.open ? " open" : ""}>
      <summary>${p.name}</summary>
      <ol>${p.pages.map((p2) => renderListItem(p2, path, normalizeLink))}
      </ol>
    </details>` : "path" in p ? html`${i > 0 && "pages" in pages[i - 1] ? html`\n  </ol>\n  <ol>` : ""}${renderListItem(
      p,
      path,
      normalizeLink
    )}` : ""
  )}
  </ol>
</nav>
<script>{${html.unsafe(
    (await rollupClient(getClientPath("sidebar-init.js"), root, path, { minify: true })).trim()
  )}}</script>`;
}
const tocSelector = "h1:not(:first-of-type), h2:first-child, :not(h1) + h2";
function findHeaders(page) {
  return Array.from(parseHtml(page.html).document.querySelectorAll(tocSelector)).map((node) => ({ label: node.textContent, href: node.firstElementChild?.getAttribute("href") })).filter((d) => !!d.label && !!d.href);
}
function renderToc(headers, label) {
  return html`<aside id="observablehq-toc" data-selector="${tocSelector}">
<nav>${headers.length > 0 ? html`
<div>${label}</div>
<ol>${headers.map(
    ({ label: label2, href }) => html`\n<li class="observablehq-secondary-link"><a href="${href}">${label2}</a></li>`
  )}
</ol>` : ""}
</nav>
</aside>`;
}
function renderListItem(page, path, normalizeLink) {
  return html`\n    <li class="observablehq-link${normalizePath(page.path) === path ? " observablehq-link-active" : ""}"><a href="${normalizeLink(relativePath(path, page.path))}">${page.name}</a></li>`;
}
function renderHead(parse, { stylesheets, staticImports, resolveImport, resolveStylesheet }, { scripts, head, root }) {
  if (parse.data?.head !== void 0)
    head = parse.data.head;
  const resolveScript = (src) => /^\w+:/.test(src) ? src : resolveImport(relativePath(root, src));
  return html`<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>${Array.from(new Set(Array.from(stylesheets, (i) => resolveStylesheet(i))), renderStylesheetPreload)}${Array.from(new Set(Array.from(stylesheets, (i) => resolveStylesheet(i))), renderStylesheet)}${Array.from(new Set(Array.from(staticImports, (i) => resolveImport(i))), renderModulePreload)}${head ? html`\n${html.unsafe(head)}` : null}${Array.from(scripts, (s) => renderScript(s, resolveScript))}`;
}
function renderScript(script, resolve) {
  return html`\n<script${script.type ? html` type="${script.type}"` : null}${script.async ? html` async` : null} src="${resolve(script.src)}"></script>`;
}
function renderStylesheet(href) {
  return html`\n<link rel="stylesheet" type="text/css" href="${href}"${/^\w+:/.test(href) ? " crossorigin" : ""}>`;
}
function renderStylesheetPreload(href) {
  return html`\n<link rel="preload" as="style" href="${href}"${/^\w+:/.test(href) ? " crossorigin" : ""}>`;
}
function renderModulePreload(href) {
  return html`\n<link rel="modulepreload" href="${href}">`;
}
function renderHeader({ header }, data) {
  if (data?.header !== void 0)
    header = data?.header;
  return header ? html`\n<header id="observablehq-header">\n${html.unsafe(header)}\n</header>` : null;
}
function renderFooter(path, options, data, normalizeLink) {
  let footer = options.footer;
  if (data?.footer !== void 0)
    footer = data?.footer;
  const link = options.pager ? findLink(path, options) : null;
  return link || footer ? html`\n<footer id="observablehq-footer">${link ? renderPager(path, link, normalizeLink) : ""}${footer ? html`\n<div>${html.unsafe(footer)}</div>` : ""}
</footer>` : null;
}
function renderPager(path, { prev, next }, normalizeLink) {
  return html`\n<nav>${prev ? renderRel(path, prev, "prev", normalizeLink) : ""}${next ? renderRel(path, next, "next", normalizeLink) : ""}</nav>`;
}
function renderRel(path, page, rel, normalizeLink) {
  return html`<a rel="${rel}" href="${normalizeLink(relativePath(path, page.path))}"><span>${page.name}</span></a>`;
}
export {
  renderPage
};
